### ISSUE TITLE
Hooks-style usage

### ISSUE THREAD

**Issue by unframework** [SupremeTechnopriest/react-idle-timer#102]
Not sure if I missed it in the docs, but has there been any discussion re: implementing hooks-based usage? I.e. allowing the library to be invoked as `useIdleTimer(...)` hook in addition to `<IdleTimer ... />`  component?

Thanks!

**Comment by SupremeTechnopriest** [comment]
Hey @unframework, 

This is the first request for a functional component which I'm actually quite surprised about.  I have had this on my mind for a while now. Perhaps I will carve out some time this weekend and make a first pass at it.

**Comment by unframework** [comment]
Yep, I thought there would have been some requests for it already (when hooks first came out, almost every library had a "hooks?" issue logged, haha). If time is tight, I can also take a stab at it, just would need to catch up with architecture/etc.

**Comment by SupremeTechnopriest** [comment]
@unframework Time this week is a bit tight, but I should be able to get to it this weekend.  Feel free to submit a PR if you like!  We can work on it together!  Otherwise ill have a new version pushed by Monday.  I don't foresee it being that challenging.

**Comment by picolloo** [comment]
I'd like to help with this one if you guys don't mind. I just have a question before addressing this, should the class component be replaced by the hook's version or just a hook wrapper solve this?

**Comment by picolloo** [comment]
I'd like to help with this one if you guys don't mind. I just have a question before addressing this, should the class component be replaced by the hook's version or just a hook wrapper solve this?

**Comment by SupremeTechnopriest** [comment]
My thoughts were to refactor the exports so you can import either like this:

```js
import { IdleTimer, useIdleTimer } from 'react-idle-timer'
```

This way you can use the class component or the hook.  I was going to write the hook from scratch and leave the class as is. Im hoping to have time this weekend to get to it.  Its been busy....

**Comment by picolloo** [comment]
I'm actually rewriting it using hooks, but the compatibility with those who use child components inside of it would be lost. Should I keep on this or try another way?

**Comment by SupremeTechnopriest** [comment]
Im imagining this will be a breaking change anyway.  I want to change the import from default to named.  Why dont you continue and submit a PR and we can go from there?  Dont forget to test :)

**Comment by unframework** [comment]
@SupremeTechnopriest this does not have to be a breaking change if the component stays as a default import like it is now, and the hook is imported by name, like so:

```
import IdleTimer, { useIdleTimer } from 'react-idle-timer'
```

Feels like less breakage would be smoother :)

Also, if I understand @picolloo correctly, the main logic can exist inside the hook, and the component implementation can invoke the hook but also contain child components/etc as with legacy usage - that makes a lot of sense to me personally.

**Comment by SupremeTechnopriest** [comment]
@unframework Yeah I had that written out at first... but I'm slowly phasing out default exports from all my work for a number of reasons.  I figured if we were going to bump to v5 we might as well make any breaking changes now.  Im ok with either approach.  Lets get the code hashed out and then we can decide how we want it exported.

**Comment by picolloo** [comment]
I rewrote the component using hooks and now I'm working on the tests. I was wondering if the usage would be something like `const { isIdle, isActive, isAction } = useIdleTimer()`.

I have to figure out how I'm supposed to pass many variables into the hook, it might be an object so the arguments are named?!

**Comment by picolloo** [comment]
I'm going to open PR for you to see if I'm doing this correctly. I haven't yet refactored the tests to work with a hook implementation, because I want some feedback before continue.

**Comment by unframework** [comment]
Thanks @picolloo for taking initiative on this!

Passing an options object into hook makes sense - would be ideal if properties match existing prop naming for component usage.

In terms of what is being returned - it could be good to have boolean flags returned as described above, but also any instance methods currently available should be exposed too (e.g. reset, etc). Basically, a 1-to-1 naming match would be nice to maintain. The new component implementation can then just put the same object into useImperativeHandle and legacy code would be able to keep working the same without breaking.

Speaking of the boolean flag - what would "isAction" represent in the flags?

**Comment by picolloo** [comment]
Oh thank you @unframework, I really wanted to help out some open-source library so I think here is my chance.

About the _isAction_ flag, I was thinking it would great to follow some pattern on the flags that the library is exposing, so a boolean for the action could be handled for some _useEffect_ on the client's application.

I have to say that I have a little hard time understanding what the _debounced_ and _throttled_ functions are supposed to do. With flags instead of callbacks, what would be debounced or throttled?

**Comment by SupremeTechnopriest** [comment]
Hey @picolloo!  

Thanks for taking charge on this. So for your first question, definitely an object as options.  You should use the same configuration as the class component.   It should return another object with the public API (the boolean callbacks and the methods like `reset`, `pause`, etc).  

Debounce and throttle are there to limit the amount of times an action is triggered within the given window of time.  For example, if you are listening to the event `mousemove`  you might not want to trigger the action for every pixel it traverses.  You probably would want to throttle or debounce the actions to once every 500ms or so.  This way you aren't spamming events.  Here is the difference between debounce and throttle (taken from stack overflow):

- **Throttling** will delay executing a function. It will reduce the notifications of an event that fires multiple times.

- **Debouncing** will bunch a series of sequential calls to a function into a single call to that function. It ensures that one notification is made for an event that fires multiple times.

I

**Comment by picolloo** [comment]
Oh I see @SupremeTechnopriest, I got this explanation and I'm feeling more confident with the refactor. Now I'm working on the hook's tests, I'm using _@testing-library/react-hooks_ for this.

**Comment by picolloo** [comment]
I have been trying to make this hook version work as the class one, but I'm doing something wrong that makes the events being fired at every action and never get idle. I'll push my latest version and keep looking for the error.

**Comment by SupremeTechnopriest** [comment]
Hey @picolloo!  How are you making out here?  If you need a hand I can hop in this weekend and take a look!

**Comment by picolloo** [comment]
Hi @SupremeTechnopriest, I have to say that I haven't been working on this lately. When I was writing up the tests I realized that I did not take time to fully understand the tool and started the refactor.

I made a mistake that makes the _idle_ state never being fired, and the _action_ to be fired multiple times in a short period of time. I've spent a while trying to find where I messed up but with no success. I'd be glad to see how it should work with hooks but I think that's my version will not be that helpful.

**Comment by SupremeTechnopriest** [comment]
Ok no problem!  I will take a look this weekend and see what's going on with that.  I will keep you posted.

**Comment by lavebug** [comment]
Any update on that or anyone has a link to a post to convert to using hooks or functional component?

**Comment by SupremeTechnopriest** [comment]
@lavebug I've been slowly working on this in my free time (which I have not had much of lately).  I have taken over where @picolloo left off. I need to finish writing a new test suite for the hook.  I'm trying to reuse as many test assertions as I can because I really don't want to have to maintain two suites. Im about to be moving to a new house at the end of the month and based on my current workload I would say I wont have this ready until mid June, unless someone wants to tag in and help out with testing.

**Comment by picolloo** [comment]
@SupremeTechnopriest do you have any branch that contains the progress you made so far?

**Comment by SupremeTechnopriest** [comment]
@picolloo I do, I can push it up this evening.  Testing hooks is a miserable experience.....

**Comment by SupremeTechnopriest** [comment]
@picolloo Branch [hooks](https://github.com/SupremeTechnopriest/react-idle-timer/tree/hooks) pushed up.  All I managed to get done so far was import your changes, abstract some commonly used functions into a utility file and start a test file for the hooks.  Im hoping to be able to share the test assertions between both the component and the hook, but I'm not sure if that is going to be possible due to how vastly different they are implemented.  I haven't complete ruled it out, but my first swing at it was not successful.  I decided to create a new test file for the hooks to try and get the ball moving.  

If anyone wants to help out feel free to throw down.  Otherwise I will get to it when I can give it the time it deserves.

**Comment by beamercola** [comment]
@SupremeTechnopriest cool! Is this somewhat functional or WIP?

**Comment by SupremeTechnopriest** [comment]
@beamercola This is definitely a work in progress.  I wouldn't use it in production until its fully tested.  I'm not sure if it even works at all yet.  I write this component test driven.  I don't think I've actually put it in a react app since I first created it.  We sold the company that I wrote this component for and I haven't needed it since.

**Comment by SupremeTechnopriest** [comment]
@beamercola @picolloo @lavebug @unframework Hey All!  Sorry this took so long to get out the door, but its finally here and fully tested.  `useIdleTimer` published to npm as version `4.3.0`.  Let me know if you experience any issues, otherwise, enjoy!
